### PR TITLE
Update PUDL from Python 3.10 to 3.11

### PR DESCRIPTION
**Pull request by zaneselvans**
# PR Overview

Dagster [supports Python 3.11 as of v1.2.2](https://github.com/dagster-io/dagster/releases/tag/1.2.2).

This PR updates our working Python version to 3.11 in various linters and environment specifications.

There were also some deprecations in the `importlib.resources` library that were causing breakage. They got more strict about what types `importlib.resources.path()` could accept, and also that function was deprecated in favor of `importlib.resources.as_file()` so I switched everything over to the new syntax.

I was able to run the `etl_full_no_cems` job successfully on my laptop.

# PR Checklist

- [x] Merge the most recent version of the branch you are merging into (probably `dev`).
- [x] All CI checks are passing. [Run tests locally to debug failures](https://catalystcoop-pudl.readthedocs.io/en/latest/dev/testing.html#running-tests-with-tox)
- [x] Wait until #2104 has been merged into `dev` and the dust has settled before merging this into `dev`.

### CODE REVIEW

**Comment by zaneselvans** [comment]
The docs build is failing, but I don't understand what this error means:
```
Warning, treated as error:
Unknown type: placeholder
ci: exit 2 (17.33 seconds) /home/runner/work/pudl/pudl> sphinx-build -W -b html docs docs/_build/html pid=4813
.pkg: _exit> python /home/runner/micromamba-root/envs/pudl-test/lib/python3.11/site-packages/pyproject_api/_backend.py True setuptools.build_meta
  ci: FAIL code 2 (165.81=setup[115.21]+cmd[0.11,10.70,0.30,0.27,5.07,0.63,0.37,0.77,0.34,3.86,7.67,0.98,0.00,1.23,0.96,17.33] seconds)
  evaluation failed :( (165.88 seconds)
```

**Comment by zaneselvans** [comment]
I **think** the real error up there is "Unknown type: placeholder" which seems totally cryptic.

When I run the sphinx docs build outside of Tox, it is successful:

```bash
rm -rf docs/_build && sphinx-build -b html docs docs/_build/html
```

**Comment by zaneselvans** [comment]
Ooof, nice sleuthing! This is what I get for trying to change things after midnight.

Would you mind going ahead and pushing the changes? I'm going to be on some other calls and stuff for a while now.

**Comment by jdangerx** [comment]
It's done! Opted out of the `tox.ini` changes since I don't know what posargs might get pushed into `tox` during CI. When debugging in the future, we can just put the `-P` in by hand or something.

File: devtools/materialize_asset.py [hunk 1]
```diff
@@ -24,28 +24,26 @@ def main(asset_id):
...
-                selection=AssetSelection.keys(asset_id).upstream(),
-                config={
-                    "resources": {
-                        "dataset_settings": {
-                            "config": EtlSettings.from_yaml(
-                                importlib.resources.path(
-                                    "pudl.package_data.settings", "etl_fast.yml"
-                                )
-                            ).datasets.dict()
+    pkg_source = files("pudl.package_data.settings").joinpath("etl_fast.yml")
```
Comment (jdangerx) [inline devtools/materialize_asset.py:27]: Neat! Wonder why they deprecated `importlib.resources.path` though 🤷

File: src/pudl/workspace/setup.py [hunk 2]
```diff
@@ -1,5 +1,5 @@
 """Tools for setting up and managing PUDL workspaces."""
-import importlib
+import importlib.resources as ir
```
Comment (jdangerx) [inline src/pudl/workspace/setup.py:2]: Is this a common alias for `importlib.resources`?

**Review by jdangerx — APPROVED** [review]
Looks good to me! All the questions I have are clarifications/for my edification only.

File: src/pudl/workspace/setup.py (hunk 2 above)
Comment (zaneselvans) [reply src/pudl/workspace/setup.py:2]: No. ☹️ 

`importlib.resources` is sooo looong, but the method names are so short and generic that they collide with other things `files` and `as_file`. I guess I should really just use the full path everywhere for clarity.

File: devtools/materialize_asset.py (hunk 1 above)
Comment (zaneselvans) [reply devtools/materialize_asset.py:27]: It definitely was not clear to me from the docs. But `importlib.resources` has always seemed mysterious to me.

**Comment by zaneselvans** [comment]
This is ready to go and I will keep it up to date with `dev` but @catalyst-cooperative/inframundo chatted about it and decided to put off merging it in until after we'd gotten the initial Dagster merge settled out on `dev`.

**Pull request merged**